### PR TITLE
Pilotwings64: Add airplanes and some SPTH parsing.

### PR DESCRIPTION
**Pull request by PFedak**
Use two derivatives of the spline (the game uses one) to make
the logic less dependent on a monotonic timer.

Only the plane SPTH objects are parsed now - the others have mismatched
numbers of keyframes, so might be interpreted differently.

### CODE REVIEW

File: src/Pilotwings64/Scenes.ts [hunk 1]
```diff
@@ -2158,7 +2351,14 @@ async function fetchDataHolder(dataFetcher: DataFetcher, device: GfxDevice): Pro
             dataHolder.uvtr.push(parseUVTR(file));
         } else if (file.type === 'UVLV') {
             dataHolder.uvlv.push(parseUVLV(file));
+        } else if (file.type === 'SPTH') {
+            const spth = parseSPTH(file);
+            if (spth !== null) {
+                dataHolder.splineData.set(userFileCounter, spth);
+            }
         }
+        if (file.type.substring(0, 2) !== 'UV')
```
Comment (magcius) [inline src/Pilotwings64/Scenes.ts]: `!file.type.startsWith('UV')`

File: src/Pilotwings64/Scenes.ts [hunk 2]
```diff
@@ -793,6 +794,115 @@ function parseUVMD(file: Pilotwings64FSFile): UVMD {
...
+    }
+    return { xSpline, ySpline, zSpline };
+}
+
+function getTwoDerivativeHermite(p0: number, p1: number, s0: number, s1: number, t: number): AnimationTrackSample {
+    const cf0 = (p0 * 2) + (p1 * -2) + (s0 * 1) + (s1 * 1);
+    const cf1 = (p0 * -3) + (p1 * 3) + (s0 * -2) + (s1 * -1);
+    const cf2 = (p0 * 0) + (p1 * 0) + (s0 * 1) + (s1 * 0);
+    const cf3 = (p0 * 1) + (p1 * 0) + (s0 * 0) + (s1 * 0);
+    return {
```
Comment (magcius) [inline src/Pilotwings64/Scenes.ts]: This is going to generate GC garbage every frame. I typically do something like this:

```ts
class AnimationTrackSample {
    public pos = 0;
    public vel = 0;
    public acc = 0;
}

function getTwoDerivativeHermite(dst: AnimationTrackSample, p0: number, ...): void {
    dst.pos = getPointCubic(cf0, cf1, cf2, cf3, t);
}

const animSampleScratch = nArray(3, () => new AnimationTrackSample());
sampleAnimationTrackHermite(animSampleScratch[0], this.spline.xSpline, t);
```

File: src/Pilotwings64/Scenes.ts [hunk 3]
```diff
@@ -793,6 +794,115 @@ function parseUVMD(file: Pilotwings64FSFile): UVMD {
...
+    const zSpline: HermiteAnimationKeyframe[] = [];
+
+    for (let i = 0; i < file.chunks.length; i++) {
+        const tag = file.chunks[i].tag;
+        const view = file.chunks[i].buffer.createDataView();
+        const pointCount = view.getUint32(0x00);
+
+        let setValues = false;
+        let currSpline: HermiteAnimationKeyframe[];
+        if (tag === 'SCPH' || tag === 'SCPX') {
```
Comment (magcius) [inline src/Pilotwings64/Scenes.ts]: "heading", "pitch", "roll"?

File: src/Pilotwings64/Scenes.ts [hunk 4]
```diff
@@ -793,6 +794,115 @@ function parseUVMD(file: Pilotwings64FSFile): UVMD {
...
+        } else if (tag === 'SCPR' || tag === 'SCPZ') {
+            currSpline = zSpline;
+        } else {
+            assert(tag === 'SCP#' && i == file.chunks.length - 1);
+            break
+        }
+        if (tag === 'SCPX' || tag === 'SCPY' || tag === 'SCPZ')
+            setValues = true;
+
+        if (currSpline.length != pointCount && currSpline.length != 0) {
```
Comment (magcius) [inline src/Pilotwings64/Scenes.ts]: !==

File: src/Pilotwings64/Scenes.ts (hunk 3 above)
Comment (PFedak) [reply src/Pilotwings64/Scenes.ts]: i made that interpretation the default, so the plane just uses them differently

File: src/Pilotwings64/Scenes.ts (hunk 2 above)
Comment (PFedak) [reply src/Pilotwings64/Scenes.ts]: Done, hopefully I'll be able to remember this now...

File: src/Pilotwings64/Scenes.ts (hunk 3 above)
Comment (magcius) [reply src/Pilotwings64/Scenes.ts]: I was just guessing at expansions for the names, not trying to have any authority here lol.

**Pull request merged**